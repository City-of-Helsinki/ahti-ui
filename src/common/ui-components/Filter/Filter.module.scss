%button-shared {
  margin: 0.5rem;
  border-style: solid;
  border-width: 2px;
  border-color: var(--hds-ui-color-black);
  padding: 0.5rem 1rem;
  font-size: var(--hds-text-md);
}

%blue-button {
  @extend %button-shared;

  background: var(--hds-brand-color-coat-of-arms-blue);
  color: var(--hds-ui-color-white);
}

%red-button {
  @extend %button-shared;

  background: var(--hds-ui-color-error);
  color: var(--hds-ui-color-white);
}

.container {
  background-color: var(--hds-ui-color-white);
  padding: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.headerContainer {
  display: flex;
  flex-direction: row;
  justify-content: center;
  padding-bottom: 1rem;
}

.textContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.icons {
  width: 24px;
  height: 24px;
}

.iconsCheck {
  margin-right: 0.25rem;
}

.filterTitle {
  font-size: var(--hds-text-lg);
}

.filterType {
  font-size: var(--hds-text-sm);
}

.optionContainer {
  margin-top: 1rem;
  justify-content: flex-start;
  overflow-y: scroll;
}

.filterOption {
  @extend %button-shared;
}

.activeButton {
  @extend %blue-button;
}

.filterOptionContainer {
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: var(--hds-ui-color-black-10);
  padding: 1rem;
}

.footerContainer {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  border-top-width: 1px;
  border-top-style: solid;
  border-top-color: var(--hds-ui-color-black-10);
}

%footer-button-shared {
  @extend %button-shared;

  border: none;
}

.clearButton {
  @extend %footer-button-shared;
}

.showButton {
  @extend %footer-button-shared;
  @extend %blue-button;
}

.showButtonDisabled {
  @extend %footer-button-shared;
  @extend %red-button;
}
