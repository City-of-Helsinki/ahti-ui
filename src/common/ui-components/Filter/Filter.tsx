import React, { useState } from 'react';
import { IconCheck, IconClose } from 'hds-react/lib';
import classNames from 'classnames/bind';
import styles from './Filter.module.scss';

const cx = classNames.bind(styles);

export interface FilterItem<T> {
  readonly name: string;
  readonly key: T;
}

export interface FilterOptions<T> {
  readonly type: string;
  readonly items: FilterItem<T>[];
}

export interface FilterProps<T> {
  readonly title?: string;
  readonly show?: string;
  readonly noMatches?: string;
  readonly clear?: string;
  readonly className?: string;
  readonly type: string;
  readonly options: FilterOptions<T>[];
  onShow(selectedFilters: T[]): void;
  onClose(): void;
  countMatches(selectedFilters: T[]): number;
}

export type FilterI<T = any> = React.FC<FilterProps<T>>;

const Filter: FilterI = ({
  title = 'Filters',
  show = 'Show',
  noMatches = 'No matches',
  clear = 'Clear',
  className,
  type,
  options,
  onShow,
  onClose,
  countMatches
}) => {
  const [selected, setSelected] = useState<Set<any>>(new Set());

  const onToggle = (toggleItem: any) => {
    const newSelection = new Set(selected);
    if (selected.has(toggleItem)) {
      newSelection.delete(toggleItem);
    } else {
      newSelection.add(toggleItem);
    }
    setSelected(newSelection);
  };

  const showButton = () => {
    const matches = countMatches(Array.from(selected));
    return (
      <button
        className={matches > 0 ? styles.showButton : styles.showButtonDisabled}
        onClick={() => onShow(Array.from(selected))}
        disabled={matches === 0}
      >
        {matches > 0 ? `${show} ${matches}` : noMatches}
      </button>
    );
  };

  return (
    <div className={cx(styles.container, className)}>
      <div>
        <div className={styles.headerContainer}>
          <button onClick={() => onClose()} className={styles.closeButton}>
            <IconClose className={styles.icons} />
          </button>
          <div className={styles.textContainer}>
            <div className={styles.filterTitle}>{title}</div>
            <div className={styles.filterType}>{type}</div>
          </div>
        </div>
        <div className={styles.optionContainer}>
          {options.map((option, id) => (
            <div className={styles.filterOptionContainer} key={id}>
              <div>{option.type}</div>
              <div>
                {option.items.map((item, id) => (
                  <button
                    key={id}
                    onClick={() => onToggle(item.key)}
                    className={cx(styles.filterOption, {
                      activeButton: selected.has(item.key)
                    })}
                  >
                    {selected.has(item.key) && (
                      <IconCheck
                        className={cx(styles.icons, styles.iconsCheck)}
                      />
                    )}
                    {item.name}
                  </button>
                ))}
              </div>
            </div>
          ))}
        </div>
      </div>
      <div className={styles.footerContainer}>
        <button
          className={styles.clearButton}
          onClick={() => setSelected(new Set())}
        >
          {clear}
        </button>
        {showButton()}
      </div>
    </div>
  );
};

export default Filter;
