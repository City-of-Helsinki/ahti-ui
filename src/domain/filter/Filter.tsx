import React, { useCallback, useEffect, useState } from 'react';
import styles from './Filter.module.scss';
import { ReactComponent as Checkmark } from '../../assets/icons/checkmark.svg';

export interface FilterOptions<T> {
  type: string;
  items: T[];
}

export interface FilterProps<T> {
  readonly options: FilterOptions<T>[];
  onShow(selectedFilters: T[]): void;
  countMatches(selectedFilters: T[]): number;
}

function Filter<T>({ options, onShow, countMatches }: FilterProps<T>) {
  const [selected, setSelected] = useState<Set<T>>(new Set());
  const onToggle = (toggleItem: T) => {
    const newSelection = new Set(selected);
    if (selected.has(toggleItem)) {
      newSelection.delete(toggleItem);
    } else {
      newSelection.add(toggleItem);
    }
    setSelected(newSelection);
  };
  const showButton = () => {
    const matches = countMatches(Array.from(selected));
    return (
      <button
        className={matches > 0 ? styles.showButton : styles.showButtonDisabled}
        onClick={() => onShow(Array.from(selected))}
        disabled={matches === 0}
      >
        {matches > 0 ? `Näytä ${matches}` : `Ei osumia`}
      </button>
    );
  };
  return (
    <div className={styles.container}>
      <div className={styles.headerContainer}>
        <div>{'Suodattimet'}</div>
        <div>{'Saaret'}</div>
      </div>
      <div className={styles.optionContainer}>
        {options.map((option, id) => (
          <div key={id}>
            <div>{option.type}</div>
            <div>
              {option.items.map((item, id) => (
                <button
                  key={id}
                  onClick={() => onToggle(item)}
                  className={
                    selected.has(item)
                      ? styles.activeButton + ' filterOption'
                      : 'filterOption'
                  }
                >
                  {selected.has(item) && <Checkmark />}
                  {item}
                </button>
              ))}
            </div>
          </div>
        ))}
      </div>
      <div className={styles.footerContainer}>
        <button
          className={styles.clearButton}
          onClick={() => setSelected(new Set())}
        >
          {'Tyhjennä'}
        </button>
        {showButton()}
      </div>
    </div>
  );
}

export default Filter;
