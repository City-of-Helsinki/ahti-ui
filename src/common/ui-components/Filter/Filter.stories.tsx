import React from 'react';
import Filter from './Filter';

export default {
  title: 'Filter',
  component: Filter
};

export const IslandFiltering = () => {
  const filterOptions = [
    {
      type: 'Kulkuyhteydet saariin',
      items: [
        { name: 'Vesibussi', key: 'water_bus' },
        { name: 'Lautta', key: 'ferry' },
        { name: 'Silta', key: 'bridge' }
      ]
    },
    {
      type: 'Palvelut saarissa',
      items: [
        { name: 'Ravintola', key: 'restaurant' },
        { name: 'Kahvila', key: 'cafe' },
        { name: 'Sauna', key: 'sauna' }
      ]
    },
    {
      type: 'Retkeily',
      items: [
        { name: 'Telttailu sallittu', key: 'camping' },
        { name: 'Kaivo', key: 'well' }
      ]
    }
  ];

  const mockData = [
    ['restaurant'],
    ['restaurant', 'sauna'],
    ['camping'],
    ['cafe', 'water_bus']
  ];

  return (
    <Filter
      title="Suodattimet"
      show="Näytä"
      noMatches="Ei hakutuloksia"
      clear="Tyhjennä"
      type="Saaret"
      options={filterOptions}
      onShow={selectedFilters =>
        alert(`Selekoidut filtterit: ${selectedFilters}`)
      }
      onClose={() => {}}
      countMatches={selectedFilters =>
        mockData.filter(
          data => new Set([...data, ...selectedFilters]).size === data.length
        ).length
      }
    />
  );
};

IslandFiltering.story = {
  name: 'Islands'
};
