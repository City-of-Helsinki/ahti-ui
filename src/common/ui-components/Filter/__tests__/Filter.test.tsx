import React from 'react';
import Filter, { FilterOptions } from '../Filter';
import { shallow, ShallowWrapper } from 'enzyme';

describe('Filter', () => {
  const filterOptions = [
    {
      type: 'Kulkuyhteydet saariin',
      items: [
        { name: 'Vesibussi', key: 'water_bus' },
        { name: 'Lautta', key: 'ferry' },
        { name: 'Silta', key: 'bridge' }
      ]
    },
    {
      type: 'Palvelut saarissa',
      items: [
        { name: 'Ravintola', key: 'restaurant' },
        { name: 'Kahvila', key: 'cafe' },
        { name: 'Sauna', key: 'sauna' }
      ]
    },
    {
      type: 'Retkeily',
      items: [
        { name: 'Telttailu sallittu', key: 'camping' },
        { name: 'Kaivo', key: 'well' }
      ]
    }
  ];

  const mockData = [
    ['restaurant'],
    ['restaurant', 'sauna'],
    ['camping'],
    ['cafe', 'water_bus']
  ];

  let wrapper: ShallowWrapper;
  let mockOnShow: jest.Mock;
  let mockOnClose: jest.Mock;

  beforeEach(() => {
    mockOnShow = jest.fn();
    mockOnClose = jest.fn();
    wrapper = shallow(
      <Filter
        type={'Saaret'}
        options={filterOptions}
        onShow={mockOnShow}
        onClose={mockOnClose}
        countMatches={selectedFilters =>
          mockData.filter(
            data => new Set([...data, ...selectedFilters]).size === data.length
          ).length
        }
      />
    );
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('selected items have correct classname', () => {
    expect(wrapper.find('button.activeButton').length).toBe(0);
    wrapper
      .find('button.filterOption')
      .at(1)
      .simulate('click');
    expect(wrapper.find('button.activeButton').length).toBe(1);
  });

  it('callbacks are called with the current selection', () => {
    wrapper
      .find('button.filterOption')
      .at(0)
      .simulate('click');
    wrapper.find('button.showButton').simulate('click');

    expect(mockOnShow.mock.calls.length).toBe(1);
    expect(mockOnShow.mock.calls[0][0][0]).toBe('water_bus');
  });

  it('0 matches disables show', () => {
    expect(wrapper.find('button.showButton').prop('disabled')).toBeFalsy();
    wrapper
      .find('button.filterOption')
      .at(1)
      .simulate('click');
    expect(
      wrapper.find('button.showButtonDisabled').prop('disabled')
    ).toBeTruthy();
  });

  it('calls onClose when clicking cross icon', () => {
    expect(mockOnClose.mock.calls.length).toBe(0);
    wrapper.find('.closeButton').simulate('click');
    expect(mockOnClose.mock.calls.length).toBe(1);
  });
});
