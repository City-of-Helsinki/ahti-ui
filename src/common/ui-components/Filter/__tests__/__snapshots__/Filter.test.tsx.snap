// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter matches snapshot 1`] = `"<div class=\\"container\\"><div><div class=\\"headerContainer\\"><button class=\\"closeButton\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"icons\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button><div class=\\"textContainer\\"><div class=\\"filterTitle\\">Filters</div><div class=\\"filterType\\">Saaret</div></div></div><div class=\\"optionContainer\\"><div class=\\"filterOptionContainer\\"><div>Kulkuyhteydet saariin</div><div><button class=\\"filterOption\\">Vesibussi</button><button class=\\"filterOption\\">Lautta</button><button class=\\"filterOption\\">Silta</button></div></div><div class=\\"filterOptionContainer\\"><div>Palvelut saarissa</div><div><button class=\\"filterOption\\">Ravintola</button><button class=\\"filterOption\\">Kahvila</button><button class=\\"filterOption\\">Sauna</button></div></div><div class=\\"filterOptionContainer\\"><div>Retkeily</div><div><button class=\\"filterOption\\">Telttailu sallittu</button><button class=\\"filterOption\\">Kaivo</button></div></div></div></div><div class=\\"footerContainer\\"><button class=\\"clearButton\\">Clear</button><button class=\\"showButton\\">Show 4</button></div></div>"`;
